provider/azure/instance.go

```diff
@@ -184,50 +182,110 @@ func (inst *azureInstance) Addresses(ctx context.ProviderCallContext) (corenetwo
 	return addresses, nil
 }
 
-// primaryNetworkAddress returns the instance's primary corenetwork.Address for
-// the internal virtual network. This address is used to identify the machine in
-// network security rules.
-func (inst *azureInstance) primaryNetworkAddress() (corenetwork.SpaceAddress, error) {
-	for _, nic := range inst.networkInterfaces {
-		if nic.IPConfigurations == nil {
+type securityGroupInfo struct {
+	resourceGroup  string
+	securityGroup  *network.SecurityGroup
+	primaryAddress corenetwork.SpaceAddress
+}
+
+// primarySecurityGroupInfo returns info for the NIC's primary corenetwork.Address
+// for the internal virtual network, and any security group on the subnet.
+// The address is used to identify the machine in network security rules.
+func primarySecurityGroupInfo(ctx stdcontext.Context, env *azureEnviron, nic network.Interface) (*securityGroupInfo, error) {
+	if nic.IPConfigurations == nil {
+		return nil, errors.NotFoundf("internal network address or security group")
+	}
+	for _, ipConfiguration := range *nic.IPConfigurations {
+		if ipConfiguration.Subnet == nil {
 			continue
 		}
-		for _, ipConfiguration := range *nic.IPConfigurations {
-			if ipConfiguration.Subnet == nil {
-				continue
-			}
-			if !to.Bool(ipConfiguration.Primary) {
-				continue
-			}
-			privateIpAddress := ipConfiguration.PrivateIPAddress
-			if privateIpAddress == nil {
-				continue
+		if !to.Bool(ipConfiguration.Primary) {
+			continue
+		}
+		privateIpAddress := ipConfiguration.PrivateIPAddress
+		if privateIpAddress == nil {
+			continue
+		}
+		securityGroup := nic.NetworkSecurityGroup
+		if securityGroup == nil {
+			idParts := strings.Split(to.String(ipConfiguration.Subnet.ID), "/")
+			lenParts := len(idParts)
+			subnetClient := network.SubnetsClient{BaseClient: env.network}
+			subnet, err := subnetClient.Get(ctx, idParts[lenParts-7], idParts[lenParts-3], idParts[lenParts-1], "networkSecurityGroup")
```

Comment:
is there a builtin parser for this?

Reply:
I wish there was, I couldn't find one and other example code seemed to split it up manually.